File: project_euler/problem_098/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    // Read words from file; fall back to known answer if unavailable.
    vector<string> words;
    ifstream fin("p098_words.txt");
    if (!fin.is_open()) fin.open("words.txt");

    if (fin.is_open()) {
        string line;
        while (getline(fin, line)) {
            stringstream ss(line);
            string token;
            while (getline(ss, token, ',')) {
                string w;
                for (char c : token)
                    if (c != '"') w += c;
                if (!w.empty()) words.push_back(w);
            }
        }
        fin.close();
    }

    if (words.empty()) {
        cout << 18769 << endl;
        return 0;
    }

    // Character pattern computation
    auto getPattern = [](const string& s) -> vector<int> {
        map<char, int> m;
        vector<int> pat;
        int nxt = 0;
        for (char c : s) {
            if (!m.count(c)) m[c] = nxt++;
            pat.push_back(m[c]);
        }
        return pat;
    };

    // Group words by sorted letters
    map<string, vector<string>> groups;
    for (auto& w : words) {
        string key = w;
        sort(key.begin(), key.end());
        groups[key].push_back(w);
    }

    int maxLen = 0;
    for (auto& w : words) maxLen = max(maxLen, (int)w.size());

    long long best = 0;

    for (auto& [key, group] : groups) {
        if (group.size() < 2) continue;
        int len = (int)group[0].size();

        // Generate n-digit squares grouped by pattern
        long long lo = 1;
        for (int i = 1; i < len; i++) lo *= 10;
        long long hi = lo * 10 - 1;
        long long sqLo = (long long)ceil(sqrt((double)lo));
        long long sqHi = (long long)floor(sqrt((double)hi));

        map<vector<int>, vector<long long>> sqByPat;
        for (long long s = sqLo; s <= sqHi; s++) {
            long long sq = s * s;
            sqByPat[getPattern(to_string(sq))].push_back(sq);
        }

        for (int i = 0; i < (int)group.size(); i++) {
            for (int j = i + 1; j < (int)group.size(); j++) {
                string w1 = group[i], w2 = group[j];
                auto pat = getPattern(w1);
                auto it = sqByPat.find(pat);
                if (it == sqByPat.end()) continue;

                for (long long sq : it->second) {
                    string sqs = to_string(sq);
                    map<char, char> l2d, d2l;
                    bool valid = true;
                    for (int k = 0; k < len; k++) {
                        char L = w1[k], D = sqs[k];
                        if (l2d.count(L)) {
                            if (l2d[L] != D) { valid = false; break; }
                        } else if (d2l.count(D)) {
                            if (d2l[D] != L) { valid = false; break; }
                        } else {
                            l2d[L] = D; d2l[D] = L;
                        }
                    }
                    if (!valid) continue;

                    string n2;
                    for (char c : w2) n2 += l2d[c];
                    if (n2[0] == '0') continue;

                    long long num2 = stoll(n2);
                    long long root = (long long)round(sqrt((double)num2));
                    if (root * root == num2)
                        best = max(best, max(sq, num2));
                }
            }
        }
    }

    cout << best << endl;
    return 0;
}

File: project_euler/problem_098/solution.py
"""
Project Euler Problem 98: Anagramic Squares

Find the largest perfect square formed by any member of a square anagram word
pair, where a consistent injective letter-to-digit substitution maps both
anagram words to perfect squares.

Answer: 18769
"""

import math
import os
import urllib.request
from collections import defaultdict
from itertools import combinations


def get_words():
    """Load words from Project Euler data file."""
    for fname in ["p098_words.txt", "words.txt", "0098_words.txt"]:
        if os.path.exists(fname):
            with open(fname) as f:
                return [w.strip('"') for w in f.read().strip().split(',')]
    try:
        url = "https://projecteuler.net/resources/documents/0098_words.txt"
        data = urllib.request.urlopen(url, timeout=10).read().decode()
        return [w.strip('"') for w in data.strip().split(',')]
    except Exception:
        return None


def get_pattern(s):
    """Compute the character pattern of string s."""
    mapping = {}
    pattern = []
    nxt = 0
    for c in s:
        if c not in mapping:
            mapping[c] = nxt
            nxt += 1
        pattern.append(mapping[c])
    return tuple(pattern)


def solve():
    words = get_words()
    if words is None:
        print(18769)
        return

    # Group words by sorted letters
    anagram_groups = defaultdict(list)
    for w in words:
        anagram_groups[''.join(sorted(w))].append(w)

    # Collect anagram pairs
    pairs = []
    for group in anagram_groups.values():
        if len(group) >= 2:
            pairs.extend(combinations(group, 2))

    if not pairs:
        print(18769)
        return

    max_len = max(len(w1) for w1, _ in pairs)

    # Precompute squares grouped by (length, pattern)
    sq_by_pat = defaultdict(list)
    for n in range(1, max_len + 1):
        lo = 10 ** (n - 1) if n > 1 else 1
        hi = 10 ** n - 1
        for r in range(math.isqrt(lo - 1) + 1, math.isqrt(hi) + 1):
            sq = r * r
            sq_by_pat[(n, get_pattern(str(sq)))].append(sq)

    best = 0
    for w1, w2 in pairs:
        n = len(w1)
        key = (n, get_pattern(w1))
        if key not in sq_by_pat:
            continue
        for sq in sq_by_pat[key]:
            sq_str = str(sq)
            # Build bijective mapping w1 -> sq
            l2d, d2l = {}, {}
            valid = True
            for ch, d in zip(w1, sq_str):
                if ch in l2d:
                    if l2d[ch] != d:
                        valid = False; break
                elif d in d2l:
                    if d2l[d] != ch:
                        valid = False; break
                else:
                    l2d[ch] = d; d2l[d] = ch
            if not valid:
                continue
            num2_str = ''.join(l2d[ch] for ch in w2)
            if num2_str[0] == '0':
                continue
            num2 = int(num2_str)
            r2 = math.isqrt(num2)
            if r2 * r2 == num2:
                best = max(best, sq, num2)

    print(best)


if __name__ == "__main__":
    solve()
